Bonsai.Core.Tests/InspectBuilderTests.cs

```diff
@@ -83,42 +83,80 @@ public void Build_CombinatorTransformError_ThrowsRuntimeException()
-            var inspectGroup = (InspectBuilder)inspectable.ElementAt(1).Value;
-            var result = inspectable.Build();
-            var visualizerElement = ExpressionBuilder.GetVisualizerElement(inspectGroup);
-            Assert.AreEqual(combinator.Value, visualizerElement.Builder);
+            ExpressionBuilder target = default;
+            var workflow = Workflow
+                .New()
+                .AppendUnit()
+                .AppendNested(
+                    input => input.AppendUnit().Do(builder => target = builder).AppendOutput(),
+                    workflow => new GroupWorkflowBuilder(workflow))
+                .ToInspectableGraph();
+            workflow.Build();
+
+            var output = workflow[workflow.Count - 1].Value;
+            var visualizerElement = ExpressionBuilder.GetVisualizerElement(output);
+            Assert.AreSame(target, visualizerElement.Builder);
         }
 
         [TestMethod]
         public void Build_PropertyMappedInspectBuilderToWorkflowOutput_ReturnVisualizerElement()
         {
-            var workflowBuilder = new WorkflowBuilder();
-            var source = workflowBuilder.Workflow.Add(new CombinatorBuilder { Combinator = new Reactive.Range() });
-            var valueSource = workflowBuilder.Workflow.Add(new CombinatorBuilder { Combinator = new IntProperty() });
-            var propertyMapping = workflowBuilder.Workflow.Add(new PropertyMappingBuilder { PropertyMappings = { new PropertyMapping { Name = "Count" } } });
-            var output = workflowBuilder.Workflow.Add(new WorkflowOutputBuilder());
-            workflowBuilder.Workflow.AddEdge(valueSource, propertyMapping, new ExpressionBuilderArgument());
-            workflowBuilder.Workflow.AddEdge(propertyMapping, source, new ExpressionBuilderArgument());
-            workflowBuilder.Workflow.AddEdge(source, output, new ExpressionBuilderArgument());
-
-            var inspectable = workflowBuilder.Workflow.ToInspectableGraph();
-            var inspectOutput = (InspectBuilder)inspectable.ElementAt(workflowBuilder.Workflow.Count - 1).Value;
-            inspectable.Build();
-            var visualizerElement = ExpressionBuilder.GetVisualizerElement(inspectOutput);
-            Assert.AreEqual(source.Value, visualizerElement.Builder);
+            ExpressionBuilder target = default;
+            var workflow = Workflow
+                .New()
+                .AppendValue(0)
+                .AppendPropertyMapping(nameof(Reactive.Range.Count))
+                .AppendCombinator(new Reactive.Range())
+                .Do(builder => target = builder)
+                .AppendOutput()
+                .ToInspectableGraph();
+            workflow.Build();
+
+            var output = workflow[workflow.Count - 1].Value;
+            var visualizerElement = ExpressionBuilder.GetVisualizerElement(output);
+            Assert.AreSame(target, visualizerElement.Builder);
+        }
+
+        [TestMethod]
```

Comment:
There is an interesting related discussion about this in https://github.com/dotnet/runtime/issues/15234 where most of my concerns about this approach were already raised. My main concern is migrating repository providers in the future, which would invalidate all URLs, but I guess this could be alleviated in a number of ways, and URLs could be easily found and updated in the future.

I would say maybe we open a separate issue for this in 2.9.0 so we can do it consistently for all tests? In the meantime I am adding backlinks as comments to these specific tests.

Reply:
Sounds good!

The way I see it is that reading tests (and wanting to know their origin) is something that happens way more often than migrating issue trackers. (This came up partially because I wanted to know where those other two tests came from in order to intentionally regress them, but their origin is just a random "Added tests" commit.)

Hopefully if we ever did migrate to something other than GitHub we'd either still have GitHub as an archive or we could mass update the URLs when we migrate.